### PR TITLE
Pass OptionSet<PlatformEventModifier> into WebKit instead of raw modifier flags in mouseDidMoveOverElement codepath

### PR DESCRIPTION
**Pull request by aprotyas**
#### 310ec15a368c09863afe46cc6179384bace485a7
<pre>
Pass OptionSet&lt;PlatformEventModifier&gt; into WebKit instead of raw modifier flags in mouseDidMoveOverElement codepath
<a href="https://bugs.webkit.org/show_bug.cgi?id=257175">https://bugs.webkit.org/show_bug.cgi?id=257175</a>
rdar://109690336

Reviewed by Aditya Keerthi.

Currently we pass event modifiers as unsigned integral values rather than
option sets into WebCore::Chrome (and then upwards into the WebKit layer).
This is because we don&apos;t want to drag in WebKit types into WebCore,
understandably so. However, this means we can&apos;t roundtrip between the UI
process and web process without translating back and forth between raw
modifier flag values and option sets.

Instead, we should:

(a) Pass OptionSet&lt;PlatformEventModifier&gt; wherever the event modifier is
required within WebCore -- i.e. In the InspectorInstrumentation path.
(b) Pass OptionSet&lt;PlatformEventModifier&gt; into our first call in the
WebKit layer, following which we can early convert to WebKit&apos;s
OptionSet&lt;WebEventModifier&gt; and pass that around inside the WebKit
layers.

This commit achieves both of the above, as well as performing a drive-by
fix of actually adding WebEventModifier.cpp to the XCode project and
moving WebKit::modifiersFromPlatformEventModifiers from
WebChromeClient.cpp into WebEventModifier.cpp, where it should reside.
The status quo seems to have just been the result of a misstep from what
was originally the right intention in 256594@main.

* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::mouseDidMoveOverElementImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::mouseDidMoveOverElement):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::mouseDidMoveOverElement):
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::mouseDidMoveOverElement):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::mouseMoved):
* Source/WebKit/Shared/WebEventModifier.cpp:
(WebKit::modifiersFromPlatformEventModifiers):

Define WebKit::modifiersFromPlatformEventModifiers that allows us to
pass a PlatformEventModifier option set into WebKit and conveniently
translate that to a WebEventModifier option set.

(WebKit::modifiersForNavigationAction):
* Source/WebKit/Shared/WebEventModifier.h:

WebCore contains the WebCore::platform function that translates from an
OptionSet&lt;WebKit::WebEventModifier&gt; to an
OptionSet&lt;WebCore::PlatformEventModifier&gt;, but the converse did not
exist. We declare WebKit::modifiersFromPlatformEventModifiers to serve
that need.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchMouseDidMoveOverElementAsynchronously):
(WebKit::WebPageProxy::mouseDidMoveOverElement):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::mouseDidMoveOverElement):
(WebKit::modifiersForNavigationAction): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::performHitTestForMouseEvent):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::mouseDidMoveOverElement):

Canonical link: <a href="https://commits.webkit.org/264519@main">https://commits.webkit.org/264519@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/681710d6ea9018fbdf71f8607dabf68be7d849a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7878 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8156 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9527 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8003 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7883 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8073 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10869 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8018 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9118 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9645 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6423 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7184 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14809 "1 flakes 112 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7546 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7315 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10700 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6338 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7115 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1880 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11324 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7532 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/Chrome.h [hunk 1]
```diff
@@ -71,6 +72,7 @@ class IntRect;
 class LocalFrame;
 class NavigationAction;
 class Page;
+enum class PlatformEventModifier : uint8_t;
```
Comment (pxlcoder) [inline Source/WebCore/page/Chrome.h]: This should be moved up near the other `enum class`.

File: Source/WebCore/page/Chrome.h [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #include "DisabledAdaptations.h"
 #include "FocusDirection.h"
 #include "HostWindow.h"
+#include "PlatformEvent.h"
```
Comment (pxlcoder) [inline Source/WebCore/page/Chrome.h]: I don't think we need the include here, since we have the forward declaration?

File: Source/WebCore/inspector/agents/InspectorDOMAgent.h [hunk 3]
```diff
@@ -32,6 +32,7 @@
 #include "EventTarget.h"
 #include "InspectorOverlay.h"
 #include "InspectorWebAgentBase.h"
+#include "PlatformEvent.h"
```
Comment (pxlcoder) [inline Source/WebCore/inspector/agents/InspectorDOMAgent.h]: Can we forward declare?

File: Source/WebKit/UIProcess/API/APIUIClient.h [hunk 4]
```diff
@@ -65,6 +65,7 @@ enum class MediaProducerMediaState : uint32_t;
 struct FontAttributes;
 struct WindowFeatures;
 using MediaProducerMediaStateFlags = OptionSet<MediaProducerMediaState>;
+enum class PlatformEventModifier : uint8_t;
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/API/APIUIClient.h]: Move up near other `enum class`.

File: Source/WebKit/UIProcess/WebPageProxy.h [hunk 5]
```diff
@@ -105,6 +105,7 @@ class IntSize;
 class LayoutPoint;
 class LayoutSize;
 class NotificationResources;
+enum class PlatformEventModifier : uint8_t;
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebPageProxy.h]: Move down near other `enum class`.

**Review by pxlcoder — COMMENTED** [review]
I think we should probably go with the alternate approach you suggested and keep using `WebEventModifier` in WebKit where possible, converting early.

This is because `WebEventModifier` is used throughout WebKit in more places, and it's strange to have differences.

File: Source/WebCore/page/Chrome.h (hunk 1 above)
Comment (aprotyas) [reply Source/WebCore/page/Chrome.h]: Fixed in https://github.com/WebKit/WebKit/commit/c362f4b3787e3a1ab3edc66da8512902bcfb5587

File: Source/WebCore/page/Chrome.h (hunk 2 above)
Comment (aprotyas) [reply Source/WebCore/page/Chrome.h]: Fixed in https://github.com/WebKit/WebKit/commit/c362f4b3787e3a1ab3edc66da8512902bcfb5587

File: Source/WebCore/inspector/agents/InspectorDOMAgent.h (hunk 3 above)
Comment (aprotyas) [reply Source/WebCore/inspector/agents/InspectorDOMAgent.h]: Don't need this anymore

File: Source/WebKit/UIProcess/API/APIUIClient.h (hunk 4 above)
Comment (aprotyas) [reply Source/WebKit/UIProcess/API/APIUIClient.h]: Don't need this anymore

File: Source/WebKit/UIProcess/WebPageProxy.h (hunk 5 above)
Comment (aprotyas) [reply Source/WebKit/UIProcess/WebPageProxy.h]: I moved it, but moved it to the wrong place 🤦‍♂️ Will fix shortly

Comment (aprotyas) [reply Source/WebKit/UIProcess/WebPageProxy.h]: Fixed in https://github.com/WebKit/WebKit/commit/e52dff05965acfedc1bef72857dda3d34a9aef8a

Comment (aprotyas) [reply Source/WebKit/UIProcess/WebPageProxy.h]: Scratch that... we don't need this forward decl at all. Removed in https://github.com/WebKit/WebKit/commit/cef350bd53f4108030d9c75e78b458f4bd5f8546

File: Source/WebKit/Shared/WebEventModifier.h [hunk 6]
```diff
@@ -41,6 +42,7 @@ enum class WebEventModifier : uint8_t {
     CapsLockKey = 1 << 4,
 };
 
+OptionSet<WebEventModifier> toWeb(OptionSet<WebCore::PlatformEventModifier>);
```
Comment (pxlcoder) [inline Source/WebKit/Shared/WebEventModifier.h]: I would probably call this `modifiersFromPlatformEventModifiers`, since we don't really use `toWeb` elsewhere. 

`from` would match `fromCorePage`/`Frame`/`Overlay` etc.

File: Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp [hunk 7]
```diff
@@ -763,16 +763,33 @@ void WebChromeClient::unavailablePluginButtonClicked(Element& element, RenderEmb
     UNUSED_PARAM(pluginUnavailabilityReason);
 }
 
-void WebChromeClient::mouseDidMoveOverElement(const HitTestResult& hitTestResult, unsigned modifierFlags, const String& toolTip, TextDirection)
+OptionSet<WebEventModifier> toWeb(OptionSet<WebCore::PlatformEventModifier> modifiers)
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp]: This should be in a new file `WebEventModifier.cpp`. It's strange to declare a method in one header and then implement in an unrelated file in WebKit.

File: Source/WebCore/page/Chrome.cpp [hunk 8]
```diff
@@ -353,9 +352,9 @@ void Chrome::mouseDidMoveOverElement(const HitTestResult& result, unsigned modif
     String toolTip;
     TextDirection toolTipDirection;
     getToolTip(result, toolTip, toolTipDirection);
-    m_client->mouseDidMoveOverElement(result, modifierFlags, toolTip, toolTipDirection);
+    m_client->mouseDidMoveOverElement(result, modifiers, toolTip, toolTipDirection);
 
-    InspectorInstrumentation::mouseDidMoveOverElement(m_page, result, modifierFlags);
+    InspectorInstrumentation::mouseDidMoveOverElement(m_page, result, modifiers.toRaw());
```
Comment (pxlcoder) [inline Source/WebCore/page/Chrome.cpp]: As discussed offline, we should try to get rid of the `toRaw` here if possible.

File: Source/WebKit/Shared/WebEventModifier.h (hunk 6 above)
Comment (aprotyas) [reply Source/WebKit/Shared/WebEventModifier.h]: I like the naming idea. Will follow suit

File: Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp (hunk 7 above)
Comment (aprotyas) [reply Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp]: I think I was following the wrong precedent set by `OptionSet<WebEventModifier> modifiersForNavigationAction(const NavigationAction&);` declared in the same header... but that's possibly because it's use is limited to the narrow scope of `WebChromeClient.cpp` where it's implemented.

I'll move out the function definition to `WebEventModifier.cpp` instead.

File: Source/WebKit/Shared/WebEventModifier.h (hunk 6 above)
Comment (aprotyas) [reply Source/WebKit/Shared/WebEventModifier.h]: Renamed in https://github.com/WebKit/WebKit/commit/1b59364ae200544534b097906f0eca943e445b7a

File: Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp (hunk 7 above)
Comment (aprotyas) [reply Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp]: So as we discovered offline, `WebEventModifier.cpp` existed in tree but was never part of the XCode project? Did a drive-by fix to rectify that and moved both `modifiersForNavigationAction` and `modifiersFromPlatformEventModifiers` into said file.

https://github.com/WebKit/WebKit/commit/1b59364ae200544534b097906f0eca943e445b7a

File: Source/WebCore/page/Chrome.cpp (hunk 8 above)
Comment (aprotyas) [reply Source/WebCore/page/Chrome.cpp]: Fixed in https://github.com/WebKit/WebKit/commit/1b59364ae200544534b097906f0eca943e445b7a

**Review by pxlcoder — APPROVED** [review]
